cli/vscodeipc/vscodeipc.go

```diff
@@ -0,0 +1,298 @@
+package vscodeipc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/google/uuid"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/xerrors"
+	"tailscale.com/tailcfg"
+
+	"github.com/coder/coder/agent"
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/codersdk"
+)
+
+// New creates a VS Code IPC client that can be used to communicate with workspaces.
+//
+// Creating this IPC was required instead of using SSH, because we're unable to get
+// connection information to display in the bottom-bar when using SSH. It's possible
+// we could jank around this (maybe by using a temporary SSH host), but that's not
+// ideal.
+//
+// This persists a single workspace connection, and lets you execute commands, check
+// for network information, and forward ports.
+//
+// The VS Code extension is located at https://github.com/coder/vscode-coder. The
+// extension downloads the slim binary from `/bin/*` and executes `coder vscodeipc`
+// which calls this function. This API must maintain backawards compatibility with
+// the extension to support prior versions of Coder.
+func New(ctx context.Context, client *codersdk.Client, agentID uuid.UUID, options *codersdk.DialWorkspaceAgentOptions) (http.Handler, io.Closer, error) {
+	if options == nil {
+		options = &codersdk.DialWorkspaceAgentOptions{}
+	}
+	// We need this to track upload and download!
+	options.EnableTrafficStats = true
+
+	agentConn, err := client.DialWorkspaceAgent(ctx, agentID, options)
+	if err != nil {
+		return nil, nil, err
+	}
+	api := &api{
+		agentConn: agentConn,
+	}
+	r := chi.NewRouter()
+	// This is to prevent unauthorized clients on the same machine from executing
+	// requests on behalf of the workspace.
+	r.Use(func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			token := r.Header.Get("Coder-Session-Token")
```

Comment:
This is similar or identical to the coder session token used by actual requests to coder, you should use a different one to avoid confusion as to what the token should be. like `Coder-Ipc-Token`